File: iOS/Sources/BeagleUI/Sources/Common/Util/Result.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Copied from Swift code base (https://github.com/apple/swift/blob/master/stdlib/public/core/Result.swift)
// to enable usage with Xcode 10.1-

/// A value that represents either a success or a failure, including an
/// associated value in each case.
public enum Result<Success, Failure: Error> {
  /// A success, storing a `Success` value.
  case success(Success)

  /// A failure, storing a `Failure` value.
  case failure(Failure)

  /// Returns a new result, mapping any success value using the given
  /// transformation.
  ///
  /// Use this method when you need to transform the value of a `Result`
  /// instance when it represents a success. The following example transforms
  /// the integer success value of a result into a string:
  ///
  ///     func getNextInteger() -> Result<Int, Error> { /* ... */ }
  ///
  ///     let integerResult = getNextInteger()
  ///     // integerResult == .success(5)
  ///     let stringResult = integerResult.map({ String($0) })
  ///     // stringResult == .success("5")
  ///
  /// - Parameter transform: A closure that takes the success value of this
  ///   instance.
  /// - Returns: A `Result` instance with the result of evaluating `transform`
  ///   as the new success value if this instance represents a success.
  public func map<NewSuccess>(
    _ transform: (Success) -> NewSuccess
  ) -> Result<NewSuccess, Failure> {
    switch self {
    case let .success(success):
      return .success(transform(success))
    case let .failure(failure):
      return .failure(failure)
    }
  }

  /// Returns a new result, mapping any failure value using the given
  /// transformation.
  ///
  /// Use this method when you need to transform the value of a `Result`
  /// instance when it represents a failure. The following example transforms
  /// the error value of a result by wrapping it in a custom `Error` type:
  ///
  ///     struct DatedError: Error {
  ///         var error: Error
  ///         var date: Date
  ///
  ///         init(_ error: Error) {
  ///             self.error = error
  ///             self.date = Date()
  ///         }
  ///     }
  ///
  ///     let result: Result<Int, Error> = // ...
  ///     // result == .failure(<error value>)
  ///     let resultWithDatedError = result.mapError({ e in DatedError(e) })
  ///     // result == .failure(DatedError(error: <error value>, date: <date>))
  ///
  /// - Parameter transform: A closure that takes the failure value of the
  ///   instance.
  /// - Returns: A `Result` instance with the result of evaluating `transform`
  ///   as the new failure value if this instance represents a failure.
  public func mapError<NewFailure>(
    _ transform: (Failure) -> NewFailure
  ) -> Result<Success, NewFailure> {
    switch self {
    case let .success(success):
      return .success(success)
    case let .failure(failure):
      return .failure(transform(failure))
    }
  }

  /// Returns a new result, mapping any success value using the given
  /// transformation and unwrapping the produced result.
  ///
  /// - Parameter transform: A closure that takes the success value of the
  ///   instance.
  /// - Returns: A `Result` instance with the result of evaluating `transform`
  ///   as the new failure value if this instance represents a failure.
  public func flatMap<NewSuccess>(
    _ transform: (Success) -> Result<NewSuccess, Failure>
  ) -> Result<NewSuccess, Failure> {
    switch self {
    case let .success(success):
      return transform(success)
    case let .failure(failure):
      return .failure(failure)
    }
  }

  /// Returns a new result, mapping any failure value using the given
  /// transformation and unwrapping the produced result.
  ///
  /// - Parameter transform: A closure that takes the failure value of the
  ///   instance.
  /// - Returns: A `Result` instance, either from the closure or the previous
  ///   `.success`.
  public func flatMapError<NewFailure>(
    _ transform: (Failure) -> Result<Success, NewFailure>
  ) -> Result<Success, NewFailure> {
    switch self {
    case let .success(success):
      return .success(success)
    case let .failure(failure):
      return transform(failure)
    }
  }

  /// Returns the success value as a throwing expression.
  ///
  /// Use this method to retrieve the value of this result if it represents a
  /// success, or to catch the value if it represents a failure.
  ///
  ///     let integerResult: Result<Int, Error> = .success(5)
  ///     do {
  ///         let value = try integerResult.get()
  ///         print("The value is \(value).")
  ///     } catch error {
  ///         print("Error retrieving the value: \(error)")
  ///     }
  ///     // Prints "The value is 5."
  ///
  /// - Returns: The success value, if the instance represents a success.
  /// - Throws: The failure value, if the instance represents a failure.
  public func get() throws -> Success {
    switch self {
    case let .success(success):
      return success
    case let .failure(failure):
      throw failure
    }
  }
}

//extension Result where Failure == Swift.Error {
//  /// Creates a new result by evaluating a throwing closure, capturing the
//  /// returned value as a success, or any thrown error as a failure.
//  ///
//  /// - Parameter body: A throwing closure to evaluate.
//  @_transparent
//  public init(catching body: () throws -> Success) {
//    do {
//      self = .success(try body())
//    } catch {
//      self = .failure(error)
//    }
//  }
//}

extension Result: Equatable where Success: Equatable, Failure: Equatable { }

extension Result: Hashable where Success: Hashable, Failure: Hashable { }

File: android/android-expression/src/main/java/br/com/zup/beagle/expression/Result.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.expression

/**
 * This class is based on flipkart-incubator/proteus's implementation
 * link @{link https://github.com/flipkart-incubator/proteus/blob/master/proteus-core/src/main/java/com/flipkart/android/proteus/toolbox/Result.java}
 * @author hernandazevedozup
 */
internal class Result(
    /**
     * Indicates the return status of the method for a given data path. The return value
     * will be `RESULT_SUCCESS` if and only if the data path exists and contains
     * a valid [com.flipkart.android.proteus.value.Value].
     */
    val resultCode: Int,
    /**
     * The value at the specified data path.
     * `value` will be null if `RESULT_CODE` != `RESULT_SUCCESS`
     */

    val value: Value
) {

    /**
     * @return true if and only if `RESULT_CODE` == `RESULT_SUCCESS`.
     */
    fun isSuccess(): Boolean = this.resultCode == RESULT_SUCCESS

    companion object {

        /**
         * Indicates that a valid [Value] was found at the specified data path.
         */
        val RESULT_SUCCESS = 0

        /**
         * Indicates that the object does not have the specified data path.
         */
        val RESULT_NO_SUCH_DATA_PATH_EXCEPTION = -1

        /**
         * Indicates that the data path specified is invalid. As an example, looking for a
         * property inside a [com.flipkart.android.proteus.value.Primitive] or [com.flipkart.android.proteus.value.Array].
         */
        val RESULT_INVALID_DATA_PATH_EXCEPTION = -2

        /**
         * Indicates that the data path prematurely led to a [com.flipkart.android.proteus.value.Null]
         */
        val RESULT_NULL_EXCEPTION = -3

        /**
         * singleton for No Such Data Path Exception.
         */
        val NO_SUCH_DATA_PATH_EXCEPTION =
            Result(RESULT_NO_SUCH_DATA_PATH_EXCEPTION, Null.INSTANCE)

        /**
         * singleton for Invalid Data Path Exception.
         */
        val INVALID_DATA_PATH_EXCEPTION =
            Result(RESULT_INVALID_DATA_PATH_EXCEPTION, Null.INSTANCE)

        /**
         * singleton for Null Exception.
         */
        val NULL_EXCEPTION = Result(RESULT_NULL_EXCEPTION, Null.INSTANCE)

        /**
         * This method return a [Result] object with `RESULT_CODE` == `RESULT_SUCCESS`
         * and `Result#value` == `value`.
         *
         * @param value The [Value] to be wrapped.
         * @return A [Result] object with with `RESULT_CODE` == `RESULT_SUCCESS`.
         */
        fun success(value: Value): Result {
            return Result(RESULT_SUCCESS, value)
        }
    }

}